### PR TITLE
Support starknet for js in sdk(fix #114)

### LINKED ISSUE #114: binding starknet to wasm

### PR DESCRIPTION
**Pull request by nick-zkp**

### CODE REVIEW

File: interface/src/json_rpc_signer.rs [hunk 1]
```diff
@@ -36,11 +62,27 @@ impl JsonRpcSigner {
...
+                    let signature = PackedEthSignature::from_hex(&s)?;
+                    let seed = signature.serialize_packed();
+                    ZkLinkSigner::new_from_seed(&seed)?
+                }
+                Layer1JsonRpcSigner::StarknetSigner(_) => {
+                    let signature = StarkECDSASignature::from_hex(&s)?;
+                    let seed = signature.signature.to_bytes_be();
+                    ZkLinkSigner::new_from_seed(&seed)?
+                }
+            }
```
Comment (Fred0327) [inline interface/src/json_rpc_signer.rs:76]: rename `self.eth_signer` to `self.l1_signer`

File: interface/src/sign_transfer.rs [hunk 2]
```diff
@@ -44,15 +51,43 @@ pub async fn sign_eth_transfer(
...
+    chain_id: &str,
+    addr: &str,
 ) -> Result<TxSignature, SignError> {
-    tx.signature = zklink_syner.sign_musig(&tx.get_bytes())?;
-    let message = tx.get_starknet_sign_msg();
-    let starknet_signature = signer.sign_message(&message)?;
+    tx.signature = zklink_signer.sign_musig(&tx.get_bytes())?;
+    let message = tx.get_starknet_sign_msg(token_symbol);
+    // #[cfg(feature = "ffi")]
+    // let message = Arc::new(message);
```
Comment (Fred0327) [inline interface/src/sign_transfer.rs:85]: remove

File: interface/src/sign_withdraw.rs [hunk 3]
```diff
@@ -1,24 +1,39 @@
 use crate::error::SignError;
+// #[cfg(feature = "ffi")]
+// use std::sync::Arc;
```
Comment (Fred0327) [inline interface/src/sign_withdraw.rs:3]: remove

File: interface/src/sign_withdraw.rs [hunk 4]
```diff
@@ -1,24 +1,39 @@
...
+    l2_source_token_symbol: &str,
+    chain_id: &str,
+    addr: &str,
 ) -> Result<TxSignature, SignError> {
     tx.sign(zklink_singer)?;
-    let message = tx.get_starknet_sign_msg();
-    let signature = signer.sign_message(&message)?;
+    let message = tx.get_starknet_sign_msg(l2_source_token_symbol);
+    // #[cfg(feature = "ffi")]
+    // let message = Arc::new(message);
```
Comment (Fred0327) [inline interface/src/sign_withdraw.rs:31]: remove

File: interface/src/signer.rs [hunk 5]
```diff
@@ -126,16 +126,23 @@ impl Signer {
...
-                sign_starknet_transfer(signer, &self.zklink_signer, tx)
-            }
+            Layer1Sginer::StarknetSigner(signer) => sign_starknet_transfer(
+                signer,
+                &self.zklink_signer,
+                tx,
+                token_symbol,
+                &chain_id.unwrap(),
+                &addr.unwrap(),
+            ),
```
Comment (Fred0327) [inline interface/src/signer.rs:145]: remove unwrap, check chain_id and addr `is_some()`

File: interface/src/signer.rs [hunk 6]
```diff
@@ -144,16 +151,23 @@ impl Signer {
...
-                sign_starknet_withdraw(signer, &self.zklink_signer, tx)
-            }
+            Layer1Sginer::StarknetSigner(signer) => sign_starknet_withdraw(
+                signer,
+                &self.zklink_signer,
+                tx,
+                l2_source_token_symbol,
+                &chain_id.unwrap(),
+                &addr.unwrap(),
+            ),
```
Comment (Fred0327) [inline interface/src/signer.rs:170]: remove unwrap, check chain_id and addr `is_some()`

File: signers/src/starknet_signer/ecdsa_signature.rs [hunk 7]
```diff
@@ -30,6 +31,20 @@ impl StarkSignature {
...
+        Self::from_bytes_be(&bytes)
+    }
+
+    pub fn from_str(r: &str, s: &str) -> Result<Self, StarkSignerError> {
+        let r = FieldElement::from_str(r)
+            .map_err(|e| StarkSignerError::InvalidSignature(e.to_string()))?;
+        let s = FieldElement::from_str(s)
+            .map_err(|e| StarkSignerError::InvalidSignature(e.to_string()))?;
+        Ok(Self { s, r })
+    }
```
Comment (Fred0327) [inline signers/src/starknet_signer/ecdsa_signature.rs:46]: rename to `from_rs_str`

File: signers/src/starknet_signer/ecdsa_signature.rs [hunk 8]
```diff
@@ -30,6 +31,20 @@ impl StarkSignature {
         hex::encode(bytes)
     }
 
+    pub fn from_hex(s: &str) -> Result<Self, StarkSignerError> {
```
Comment (Fred0327) [inline signers/src/starknet_signer/ecdsa_signature.rs:34]: rename StarkSignature to `Stark EcdsaSignature`

File: signers/src/starknet_signer/ecdsa_signature.rs [hunk 9]
```diff
@@ -93,9 +108,11 @@ impl StarkECDSASignature {
 }
 
 impl StarkECDSASignature {
-    pub fn verify(&self, msg: &[u8]) -> Result<bool, StarkSignerError> {
+    pub fn verify(&self, msg: &TypedData, addr: &str) -> Result<bool, StarkSignerError> {
+        let addr = FieldElement::from_hex_be(addr)
+            .map_err(|e| StarkSignerError::SignError(e.to_string()))?;
+        let hash = msg.get_message_hash(addr)?;
         let verifying_key = VerifyingKey::from_scalar(self.pub_key);
```
Comment (Fred0327) [inline signers/src/starknet_signer/ecdsa_signature.rs:115]: rename StarkECDSASignature to `StarkSignature`

File: signers/src/starknet_signer/pk_signer.rs [hunk 10]
```diff
@@ -73,14 +80,55 @@ mod tests {
...
-        let msg = b"hello world";
-        let signature = stark_signer.sign_message(msg).unwrap();
-        let is_ok = signature.verify(msg).unwrap();
+        let msg_hash = "0x51d5faacb1bdeb6293d52fd4be0a7c62417cb73962cdd6aff385b67239cf081";
+        let signature = stark_signer
+            .0
+            .sign(&FieldElement::from_hex_be(msg_hash).unwrap())
+            .unwrap();
+        //let pub_key = stark_signer.public_key();
+        let pub_key = "0x3b478eae5afdd35358abcc7955bba7acda3d16f4485b62f2497f78ed6bc7126";
```
Comment (Fred0327) [inline signers/src/starknet_signer/pk_signer.rs:89]: pub_keys are same or not?

File: signers/src/starknet_signer/pk_signer.rs [hunk 11]
```diff
@@ -73,14 +80,55 @@ mod tests {
...
+        println!("{:?}", msg_hash);
+        let signature = StarkECDSASignature::from_hex(sig_str).unwrap();
+        let verifying_key = VerifyingKey::from_scalar(pub_key);
+        let is_ok = verifying_key
+            .verify(
+                &FieldElement::from_hex_be(&hex::encode(msg_hash.to_bytes_be())).unwrap(),
+                &Signature {
+                    s: signature.signature.s,
+                    r: signature.signature.r,
+                },
```
Comment (Fred0327) [inline signers/src/starknet_signer/pk_signer.rs:129]: add `address` to StarkSignature, so ther are tree fileds in `StarkSignature`:
```
{
    address: xxx,
    pub_key: xxx,
    signature: StarkEcdsaSignature,
}
```

so, when verify the `StarkSignature`:
1. create typed_data with Transaction, then create msg_hash
2. verify msg_hash with pub_key

File: signers/src/starknet_signer/typed_data/mod.rs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+    pub fn get_struct_hash<M: Serialize>(
+        &self,
+        struct_type: &str,
+        data: &M,
+    ) -> Result<FieldElement, StarkSignerError> {
+        let mut types_arry = vec![];
+        let mut data_arry = vec![];
+        types_arry.push("felt".to_string());
+        let type_hash = starknet_keccak(self.encode_type(struct_type).as_bytes());
```
Comment (Fred0327) [inline signers/src/starknet_signer/typed_data/mod.rs:158]: add unit test

File: signers/src/starknet_signer/pk_signer.rs (hunk 11 above)
Comment (nick-zkp) [reply signers/src/starknet_signer/pk_signer.rs:129]: we can get account address from tx's from account.

File: interface/src/sign_transfer.rs [hunk 13]
```diff
@@ -1,8 +1,15 @@
 use crate::error::SignError;
+// #[cfg(feature = "ffi")]
+// use std::sync::Arc;
```
Comment (Fred0327) [inline interface/src/sign_transfer.rs:3]: remove

File: signers/src/starknet_signer/typed_data/mod.rs [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+use num::{BigUint, Num};
+use serde::{Deserialize, Serialize};
+
+use starknet_core::crypto::compute_hash_on_elements;
+use starknet_core::types::FieldElement;
+use starknet_core::utils::starknet_keccak;
+use std::fmt::Debug;
+use std::str::FromStr;
+// #[cfg(feature = "ffi")]
+// use std::sync::Arc;
```
Comment (Fred0327) [inline signers/src/starknet_signer/typed_data/mod.rs:14]: remove

File: signers/src/zklink_signer/pk_signer.rs [hunk 15]
```diff
@@ -12,10 +12,16 @@ use franklin_crypto::bellman::{PrimeField, PrimeFieldRepr};
 use franklin_crypto::eddsa::{PrivateKey as FLPrivateKey, PrivateKey, PublicKey, Seed};
 use sha2::{Digest, Sha256};
 use std::fmt;
+// #[cfg(feature = "ffi")]
+// use std::sync::Arc;
```
Comment (Fred0327) [inline signers/src/zklink_signer/pk_signer.rs:16]: remove

**Review by Fred0327 — APPROVED** [review]

**Pull request merged**